### PR TITLE
[PLAT-78] Cooperative threads run only if tasklets are available

### PR DESCRIPTION
**Pull request by ramizdundar**
https://hazelcast.atlassian.net/browse/PLAT-78

This PR tries to solve the above mentioned problem by doing the following:

- ~Do not start cooperative threads at the startup~
- ~If any job submitted to the Platform start the cooperative threads~
- ~Cooperative threads stay active until shutdown~
- Cooperative threads sleep if there are no available tasklets
- Cooperative threads wake up if a new job submitted

This is an alternative approach to [this PR](https://github.com/hazelcast/hazelcast/pull/18574) using semaphores. It should be either that or this should be merged.

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Request reviewers if possible

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/jet/impl/execution/TaskletExecutionService.java [hunk 1]
```diff
@@ -85,6 +87,7 @@
     private final String hzInstanceName;
     private final ILogger logger;
     private int cooperativeThreadIndex;
+    private final AtomicBoolean startCooperativeThreads = new AtomicBoolean(false);
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/execution/TaskletExecutionService.java:90]: We no longer need this - we can start the threads right away and they won't take cpu.

File: hazelcast/src/main/java/com/hazelcast/jet/impl/execution/TaskletExecutionService.java [hunk 2]
```diff
@@ -325,6 +332,8 @@ public void run() {
         // prevent lambda allocation on each iteration
         private final Consumer<TaskletTracker> runTasklet = this::runTasklet;
 
+        private final Semaphore hasTasklet = new Semaphore(1);
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/execution/TaskletExecutionService.java:335]: Doesn't really change anything, but why we start with 1 permit? Should be 0.
A better name might be `newTaskletsSemaphore`.

**Review by viliam-durina — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/jet/impl/execution/TaskletExecutionService.java [hunk 3]
```diff
@@ -350,7 +355,18 @@ public void run() {
...
                     idleCount = 0;
                 } else {
-                    idlerLocal.idle(++idleCount);
+                    if (trackers.isEmpty()) {
+                        newTaskletSemaphore.drainPermits();
+                        if (trackers.isEmpty() && !isShutdown) {
+                            try {
+                                newTaskletSemaphore.acquire();
+                            } catch (InterruptedException e) {
+                                logger.fine(e.getMessage());
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/execution/TaskletExecutionService.java:364]: I guess the message is `null`.

We should not ignore the interrupt. Jet doesn't use it, so it should never happen - in that case i guess we should log it on the error level and return.
```suggestion
                                logger.severe("Cooperative worker interrupted", e);
                                return;
```

**Review by viliam-durina — APPROVED** [review]

**Review by gurbuzali — APPROVED** [review]

**Comment by ramizdundar** [comment]
run-lab-run

**Comment by mmedenjak** [comment]
@ramizdundar I see the PR has been approved. Can we merge it and close https://github.com/hazelcast/hazelcast/pull/18574?

**Review by frant-hartm — APPROVED** [review]

**Pull request merged**